File: CubosphereToRollingcubeLevelConverter/src/main/java/kp/rollingcube/levelConverter/level/properties/PropertyValue.java
package kp.rollingcube.levelConverter.level.properties;

import lombok.NonNull;

/**
 *
 * @author Marc
 */
public final class PropertyValue
{
    private int _int;
    private float _float;
    private @NonNull String _str;
    
    private PropertyValue(int intValue, float floatValue, @NonNull String strValue)
    {
        this._int = intValue;
        this._float = floatValue;
        this._str = strValue;
    }
    
    public static @NonNull PropertyValue of() { return new PropertyValue(0, 0, ""); }
    public static @NonNull PropertyValue of(int value) { return new PropertyValue(value, 0, ""); }
    public static @NonNull PropertyValue of(float value) { return new PropertyValue(0, value, ""); }
    public static @NonNull PropertyValue of(boolean value) { return new PropertyValue(value ? 1 : 0, 0, ""); }
    public static @NonNull PropertyValue of(@NonNull String value) { return new PropertyValue(0, 0, value); }
    
    public static @NonNull PropertyValue zero() { return of(); }
    
    public static @NonNull PropertyValue fromStringAsInteger(String value) throws NumberFormatException
    {
        if(value == null)
            value = "0";
        
        return of(Integer.parseInt(value));
    }
    public static @NonNull PropertyValue fromStringAsFloat(String value) throws NumberFormatException
    {
        if(value == null)
            value = "0";
        
        return of(Float.parseFloat(value));
    }
    
    public static @NonNull PropertyValue fromStringAsBoolean(String value)
    {
        if(value == null)
            value = "false";
        
        return of(value.equalsIgnoreCase("true"));
    }
    
    public static @NonNull PropertyValue fromStringAsString(String value) { return of(value); }
    
    public final int getInteger() { return _int; }
    public final float getFloat() { return _float; }
    public final boolean getBoolean() { return _int != 0; }
    public final @NonNull String getString() { return _str; }
    public final int getEnumOrdinal() { return Math.abs(_int); }
    
    public void setInteger(int value) { _int = value; _float = 0; _str = ""; }
    public void setFloat(float value) { _int = 0; _float = value; _str = ""; }
    public void setBoolean(boolean value) { _int = value ? 1 : 0; _float = 0; _str = ""; }
    public void setString(@NonNull String value) { _int = 0; _float = 0; _str = value; }
    public void setEnumOrdinal(int value) { _int = Math.abs(value); _float = 0; _str = ""; }
    
    public @NonNull PropertyValue copy() { return new PropertyValue(_int, _float, _str); }
    
    public @NonNull PropertyValue copyFrom(@NonNull PropertyValue other)
    {
        _int = other._int;
        _float = other._float;
        _str = other._str;
        return this;
    }
    
    public @NonNull PropertyValue copyTo(@NonNull PropertyValue other)
    {
        return other.copyFrom(this);
    }
}

File: Kotlin - CubosphereToRollingcubeLevelConverter/CubosphereToRollingcubeLevelConverter/src/main/kotlin/kp/rollingcube/lc/level/properties/PropertyValue.kt
package kp.rollingcube.lc.level.properties

import kotlin.math.absoluteValue

class PropertyValue private constructor(private var int: Int, private var float: Float, private var string: String)
{
    constructor(value: Int) : this(value, 0.0f, "")
    constructor(value: Float) : this(0, value, "")
    constructor(value: Boolean) : this(if(value) 1 else 0, 0.0f, "")
    constructor(value: String) : this(0, 0.0f, value)

    var intValue: Int get() = int; set(value) { int = value; float = 0.0f; string = ""; }
    var floatValue: Float get() = float; set(value) { int = 0; float = value; string = ""; }
    var booleanValue: Boolean get() = int != 0; set(value) { int = if(value) 1 else 0; float = 0.0f; string = ""; }
    var stringValue: String get() = string; set(value) { int = 0; float = 0.0f; string = value; }
    var enumOrdinalValue: Int get() = int.absoluteValue; set(value) { int = 0.coerceAtLeast(value); float = 0.0f; string = ""; }

    val copy: PropertyValue get() = PropertyValue(int, float, string)

    fun copyFrom(other: PropertyValue)
    {
        int = other.int
        float = other.float
        string = other.string
    }
    fun copyTo(other: PropertyValue) = other.copyFrom(this)

    companion object
    {
        val ZERO: PropertyValue get() = PropertyValue(0, 0.0f, "")

        fun Int.toPropertyValue() = PropertyValue(this)
        fun Float.toPropertyValue() = PropertyValue(this)
        fun Boolean.toPropertyValue() = PropertyValue(this)
        fun String?.toPropertyValue() = PropertyValue(this ?: "")

        fun String?.toIntPropertyValue() = if(this == null) 0.toPropertyValue() else toInt().toPropertyValue()
        fun String?.toFloatPropertyValue() = if(this == null) 0.0f.toPropertyValue() else toFloat().toPropertyValue()
        fun String?.toBooleanPropertyValue() = if(this == null) false.toPropertyValue() else toBoolean().toPropertyValue()
        fun String?.toStringPropertyValue() = toPropertyValue()
    }
}